misk-hibernate/src/main/kotlin/misk/hibernate/Query.kt

```diff
@@ -1,17 +1,31 @@
 package misk.hibernate
 
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Root
+import javax.persistence.criteria.Selection
 import kotlin.reflect.KClass
 
 /** Base class for SQL queries. */
 interface Query<T> {
+  /** The first result to return. Defaults to 0. */
+  var firstResult: Int
+
   /** How many rows to return. Must be -1 or in range 1..10_000. */
   var maxRows: Int
 
+  /** Constrain a query using a path known only at runtime. */
+  fun addConstraint(path: String, operator: Operator, value: Any? = null)
+  fun addOrder(path: String, asc: Boolean)
+  fun list(session: Session, projectedPaths: List<String>): List<List<Any?>>
```

Comment:
should we group this with the list APIs below?

Reply:
I can see it going either way. The ones above are runtime constraints. Whereas the ones below are more or less compile time. But I can group by function type if we think that is better. It might be.

Reply:
I ended up changing this. They end up matching with the manual projection comments better.

Reply:
I ended up changing this. They end up matching with the manual projection comments better.